### PR TITLE
Fix groupBy regression on time extractions named __time

### LINKED ISSUE #3683: groupBy regression on time extractions named __time
This test passed in 0.9.1.1 but fails in 0.9.2-rc2 for groupBy engine v1. It works with the v2 engine. Nevertheless I'm marking this 0.9.2 since it's a regression in the v1 engine, which is still default.

```java
  @Test
  public void testGroupByTimeExtractionNamedUnderUnderTime()
  {
    GroupByQuery query = GroupByQuery
        .builder()
        .setDataSource(QueryRunnerTestHelper.dataSource)
        .setQuerySegmentSpec(QueryRunnerTestHelper.fullOnInterval)
        .setDimensions(
            Lists.newArrayList(
                new DefaultDimensionSpec("market", "market"),
                new ExtractionDimensionSpec(
                    Column.TIME_COLUMN_NAME,
                    Column.TIME_COLUMN_NAME,
                    new TimeFormatExtractionFn("EEEE", null, null, null),
                    null
                )
            )
        )
        .setAggregatorSpecs(
            Arrays.asList(
                QueryRunnerTestHelper.rowsCount,
                QueryRunnerTestHelper.indexDoubleSum
            )
        )
        .setPostAggregatorSpecs(Arrays.<PostAggregator>asList(QueryRunnerTestHelper.addRowsIndexConstant))
        .setGranularity(QueryRunnerTestHelper.allGran)
        .setDimFilter(
            new OrDimFilter(
                Arrays.<DimFilter>asList(
                    new SelectorDimFilter("market", "spot", null),
                    new SelectorDimFilter("market", "upfront", null)
                )
            )
        )
        .build();
    List<Row> expectedResults = Arrays.asList(
        GroupByQueryRunnerTestHelper.createExpectedRow(
            "1970-01-01",
            "__time",
            "Friday",
            "market",
            "spot",
            "index",
            13219.574157714844,
            "rows",
            117L,
            "addRowsIndexConstant",
            13337.574157714844
        ),
        GroupByQueryRunnerTestHelper.createExpectedRow(
            "1970-01-01",
            "__time",
            "Monday",
            "market",
            "spot",
            "index",
            13557.738830566406,
            "rows",
            117L,
            "addRowsIndexConstant",
            13675.738830566406
        ),
        GroupByQueryRunnerTestHelper.createExpectedRow(
            "1970-01-01",
            "__time",
            "Saturday",
            "market",
            "spot",
            "index",
            13493.751281738281,
            "rows",
            117L,
            "addRowsIndexConstant",
            13611.751281738281
        ),
        GroupByQueryRunnerTestHelper.createExpectedRow(
            "1970-01-01",
            "__time",
            "Sunday",
            "market",
            "spot",
            "index",
            13585.541015625,
            "rows",
            117L,
            "addRowsIndexConstant",
            13703.541015625
        ),
        GroupByQueryRunnerTestHelper.createExpectedRow(
            "1970-01-01",
            "__time",
            "Thursday",
            "market",
            "spot",
            "index",
            14279.127197265625,
            "rows",
            126L,
            "addRowsIndexConstant",
            14406.127197265625
        ),
        GroupByQueryRunnerTestHelper.createExpectedRow(
            "1970-01-01",
            "__time",
            "Tuesday",
            "market",
            "spot",
            "index",
            13199.471435546875,
            "rows",
            117L,
            "addRowsIndexConstant",
            13317.471435546875
        ),
        GroupByQueryRunnerTestHelper.createExpectedRow(
            "1970-01-01",
            "__time",
            "Wednesday",
            "market",
            "spot",
            "index",
            14271.368591308594,
            "rows",
            126L,
            "addRowsIndexConstant",
            14398.368591308594
        ),
        GroupByQueryRunnerTestHelper.createExpectedRow(
            "1970-01-01",
            "__time",
            "Friday",
            "market",
            "upfront",
            "index",
            27297.8623046875,
            "rows",
            26L,
            "addRowsIndexConstant",
            27324.8623046875
        ),
        GroupByQueryRunnerTestHelper.createExpectedRow(
            "1970-01-01",
            "__time",
            "Monday",
            "market",
            "upfront",
            "index",
            27619.58447265625,
            "rows",
            26L,
            "addRowsIndexConstant",
            27646.58447265625
        ),
        GroupByQueryRunnerTestHelper.createExpectedRow(
            "1970-01-01",
            "__time",
            "Saturday",
            "market",
            "upfront",
            "index",
            27820.83154296875,
            "rows",
            26L,
            "addRowsIndexConstant",
            27847.83154296875
        ),
        GroupByQueryRunnerTestHelper.createExpectedRow(
            "1970-01-01",
            "__time",
            "Sunday",
            "market",
            "upfront",
            "index",
            24791.223876953125,
            "rows",
            26L,
            "addRowsIndexConstant",
            24818.223876953125
        ),
        GroupByQueryRunnerTestHelper.createExpectedRow(
            "1970-01-01",
            "__time",
            "Thursday",
            "market",
            "upfront",
            "index",
            28562.748901367188,
            "rows",
            28L,
            "addRowsIndexConstant",
            28591.748901367188
        ),
        GroupByQueryRunnerTestHelper.createExpectedRow(
            "1970-01-01",
            "__time",
            "Tuesday",
            "market",
            "upfront",
            "index",
            26968.280639648438,
            "rows",
            26L,
            "addRowsIndexConstant",
            26995.280639648438
        ),
        GroupByQueryRunnerTestHelper.createExpectedRow(
            "1970-01-01",
            "__time",
            "Wednesday",
            "market",
            "upfront",
            "index",
            28985.5751953125,
            "rows",
            28L,
            "addRowsIndexConstant",
            29014.5751953125
        )
    );
    Iterable<Row> results = GroupByQueryRunnerTestHelper.runQuery(factory, runner, query);
    TestHelper.assertExpectedObjects(expectedResults, results, "");
  }
```

### PR DESCRIPTION
**Pull request by jon-wei**
Fixes #3683

### CODE REVIEW

**Comment by fjy** [comment]
👍

File: processing/src/main/java/io/druid/segment/DimensionHandlerUtil.java [hunk 1]
```diff
@@ -35,6 +36,11 @@ public static DimensionHandler getHandlerFromCapabilities(
   )
   {
     DimensionHandler handler = null;
+
+    if (dimensionName.equals(Column.TIME_COLUMN_NAME)) {
```
Comment (fjy) [inline processing/src/main/java/io/druid/segment/DimensionHandlerUtil.java]: can we add a reminder that this logic probably needs to change with new dim typing work?

**Comment by gianm** [comment]
Hmm after thinking about this some more I wonder if it's something we really want to support. Maybe having this query fail is OK, and it should actually fail with a more descriptive error. The reason is that naming output fields `__time` is problematic for features like,

- nested queries: If the inner query has granularity "hour" and also has a dimension with outputName `__time`, and the outer query filters or groups on `__time`, is that referring to the timestamp of the inner rows or is it referring to the dimension named `__time`?
- having specs: If a query has a metric with output name `__time`, and a having spec filters on `__time`, is that referring to the timestamp of the rows or to the metric named `__time`?

Currently I believe for nested queries it refers to the timestamp but for having specs it refers to the metric named `__time`. This is inconsistent and weird and I think there's the best answer is to disallow dimensions or metrics with output names of `__time`.

**Comment by gianm** [comment]
You can see all of these issues come from the fact that groupBy results have anonymous timestamps in addition to actual named fields coming from dims and metrics and postaggregators. "Higher level" things like having specs and nested queries can access the timestamp by calling it `__time`, but that ends up shadowing any similarly named field from a dim or metric or postaggregator.

As a separate but related issue, I think nothing is really stopping anyone from defining a dimension and metric with the same output name. We should detect and fail on that at the constructor level, since otherwise one will clobber the other.

**Comment by gianm** [comment]
Maybe since this used to work, we can take the attitude that for 0.9.x that the behavior is that you are allowed to have a dimension or metric with name `__time`, and it will be returned to you just fine in the query result, _but_ "higher level" Druid constructs like having specs and outer queries will not be able to see it. Instead, they'll see the anonymous timestamp of the row. I think this matches how 0.9.1.1 worked.

In 0.10.0 I'd like to make it so naming something `__time` just fails outright.

**Comment by gianm** [comment]
I suggested an alternate fix in #3685.

I also raised #3687 for discussing what to do about the remaining issues with `__time` and groupBy, and #3686 for 0.9.3 with validation of the "separate but related issue" that dimensions/metrics/postaggs could collide.

**Comment by jon-wei** [comment]
Closing in favor of #3685

**Pull request closed without merging**